torchtune/datasets/_preference.py

```diff
@@ -10,64 +10,98 @@
+    the dataset to have "chosen" and "rejected" model responses. At a high level, this
+    class will load the data from source and apply the following pre-processing steps
+    when a sample is retrieved:
+
+    1. Dataset-specific transform. This is typically unique to each dataset and extracts
+       the necessary prompt and chosen/rejected columns into torchtune's :class:`~torchtune.data.Message`
+       format, a standardized API for all model tokenizers.
+    2. If specified, apply a prompt template for the task you are fine-tuning for.
+    3. Tokenization
+
+
+    All datasets are formatted into a list of :class:`~torchtune.data.Message`
+    because preference datasets can be considered as chosen and rejected "conversations"
+    with the model, or AI assistant. Thus, we can standardize all text content as messages
+    in a conversation assigned to a role:
+
+    - ``"user"`` messages contain the input prompt into the model
+    - ``"assistant"`` messages are the response of the model and what you actually want
+      to train for and compute loss directly against
+
+    The :class:`~torchtune.data.Message` forms the core data unit that all tokenizer
+    APIs expect. The key component of this class that ensures any dataset is transformed
+    into this format is the ``message_transform``. This is a callable class that takes
+    in a sample dictionary - typically a single row from the source dataset - that
+    processes the sample in any configurable way to output a list of messages::
+
+        [
+            Message(
+                role=<system|user|assistant|ipython>,
+                content=<message>,
+            ),
+            ...
+        ]
 
-    The general flow from loading a sample to tokenized prompt is:
-    load sample -> apply transform -> format into template -> tokenize
+    For any custom dataset, use the ``message_transform`` to contain all pre-processing to
+    return the list of messages.
 
-    If the column/key names differ from the expected names in the :class:`~torchtune.data.InstructTemplate`,
-    then the ``column_map`` argument can be used to provide this mapping.
 
     Args:
-        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
         source (str): path to dataset repository on Hugging Face. For local datasets,
             define source as the data file type (e.g. "json", "csv", "text") and pass
-            in the filepath in ``data_files``. See Hugging Face's ``load_dataset``
-            (https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path)
+            in the filepath in ``data_files``. See `Hugging Face's
+            <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path>`_
+            ``load_dataset`` for more details.
+        message_transform (Transform): callable that keys into the desired fields in the sample
```

Comment:
pretty tangential point, but the behaviour in this docstring isn't enforced by the ``Transform`` protocol itself right? In that the transform should always have `def __call__(...) -> List[Message]:`